mono/metadata/handle.h

```diff
@@ -24,6 +24,36 @@
 #include <mono/utils/checked-build.h>
 #include <mono/metadata/class-internals.h>
 
+// Type-safe handles are a struct with a pointer to pointer.
+// The only operations allowed on them are the functions/macros in this file, and assignment
+// from same handle type to same handle type.
+//
+// Type-unsafe handles are a pointer to a struct with a pointer.
+// Besides the type-safe operations, these can also be:
+//  1. compared to NULL, instead of only MONO_HANDLE_IS_NULL
+//  2. assigned from NULL, instead of only a handle
+//  3. MONO_HANDLE_NEW (T) from anything, instead of only a T*
+//  4. MONO_HANDLE_CAST from anything, instead of only another handle type
+//  5. assigned from any void*, at least in C
+//  6. Cast from any handle type to any handle type, without using MONO_HANDLE_CAST.
+//  7. Cast from any handle type to any pointer type and vice versa, such as incorrect unboxing.
+//  8. mono_object_class (handle), instead of mono_handle_class
+//
+// None of those operations were likely intended.
+// Completely incidentally, the type-safe system is a little more expression-oriented than
+// the statement-based type-unsafe system, and might be valid C++ where type-unsafe is not (untested).
+
+// FIXME Do this only on checked builds? Or certain architectures?
+// There is not runtime cost.
+// NOTE: Running this code depends on the ABI to pass a struct
+// with a pointer the same as a pointer. This is tied in with
+// marshaling. If this is not the case, turn off type-safety, perhaps per-OS per-CPU.
+#if !defined (TARGET_X86) || defined (HOST_WIN32) || defined (HOST_DARWIN)
```

Comment:
ABIs should be read up on to confirm the Linux divergence. Or look at small cod e samples. Or confirm this was the problem, etc.